hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -213,513 +175,413 @@ class HandlerLowering(paths: HandlerPaths, opt: EffectHandlers)(using TL, Raise,
-
-  def partitionBlock(blk: Block, inclEntryPoint: Bool, labelIds: Map[Symbol, (StateId, StateId)] = Map.empty): Ls[BlockState] =
-    // for readability :)
-    case class PartRet(head: Block, states: Ls[BlockState])
-    
-    var resumptionPoints: List[StateId] = List.empty
+  object EffectfulResult:
+    def unapply(r: Result) = r match
+      case c: Call if c.mayRaiseEffects => S(r)
+      case _: Instantiate => S(r)
+      case _ => N
+  
+  private def partitionBlock(blk: Block)(using h: FunctionCtx): PartitionedBlock =
+    val result = mutable.HashMap.empty[StateId, BlockPartition]
+    val allocId = new IdAllocator()
+    var containsCall = false
 
-    // * returns (truncated input block, child block states)
     // * blk: The block to transform
+    // * partitioned: whether we are already in a partitioned state
+    // *              if we are not partitioned, we do not need to jump to afterEnd,
+    // *              this is because we are still in the original block, which shares
+    // *              the same code path.
     // * labelIds: maps label IDs to the state at the start of the label and the state after the label
-    // * afterEnd: what state End should jump to, if at all 
-    // TODO: don't split within Match, Begin and Labels when not needed, ideally keep it intact.
-    // Need careful analysis for this.
-    def go(blk: Block)(implicit labelIds: Map[Symbol, (StateId, StateId)], afterEnd: Option[StateId]): PartRet =
+    // * afterEnd: what state End should jump to, if at all
+    def go(blk: Block)(using labelIds: Map[Symbol, (LazyId, LazyId)], afterEnd: Option[LazyId], partitioned: Bool): Block = boundary:
+      // First check if the current block contain any non trivial call, if so we need a partition
+
+      def forceId(blk: Block, resumable: Bool): StateId = blk match
+        case StateTransition(uid) =>
+          if !result(uid).resumable && resumable then
+            result(uid) = BlockPartition(result(uid).blk, true)
+          uid
+        case _ =>
+          val id = allocId()
+          result(id) = BlockPartition(blk, resumable)
+          id
+
+      // sym: the local that stores the result
+      def doNewEffectPartition(res: Result, rst: Block) =
+        val stateId = forceId(go(rst)(using partitioned = true), true)
+        val newBlock = blockBuilder
+          .assignFieldN(paths.runtimePath, paths.resumeValueIdent, res)
+          .ifthen(
+            paths.resumeValue,
+            Case.Cls(paths.effectSigSym, paths.effectSigPath),
+            Unwind(stateId, res.toLoc.fold(unit)(locToStr(_)))
+          )
+          .rest(StateTransition(stateId))
+        boundary.break(newBlock)
+      class RestLazyId(rst: Block) extends LazyId:
+        def getImpl: StateId = forceId(go(rst)(using partitioned = true), false)
+        def transitionSoft: Block = transitionOrBlk(go(rst))
+
+      val nonTrivialBlockChecker = new BlockDataTransformer(SymbolSubst()):
```

Comment:
It appears the return result of `applyBlock` is never used. Why not make it a `BlockDataTraverser`, which is more efficient?